### PR TITLE
Add $HISTORY variable

### PR DESCRIPTION
**Pull request by samuelfekete**
Add `$HISTORY` magic variable that can be placed before other magic variables to make them refer to any previous test, e.g.  `$HISTORY['test name'].$LOCATION`.
https://github.com/cdent/gabbi/issues/149

Also added `$URL`, because `$HISTORY...$LAST_URL` would be too confusing.

### CODE REVIEW

**Comment by samuelfekete** [comment]
I think this would break the gabbihtml handler (https://github.com/cdent/gabbihtml/blob/master/gabbihtml/handler.py#L97), which uses `gen_replacer()` and not `replacer()` and it assumes that only the last response will be handled.

**Comment by cdent** [comment]
> I think this would break the gabbihtml handler

That's not a problem, it's not released or really properly done yet. I'll go back through history and try to remind myself why gen_replacer (probably an early optimization explicitly for the html handler) was created just to make sure some edge case isn't getting messed up.

I'll need to give this a closer look later in the day, but a superficial read through it looks sane and workable. Nicely done.

We'll of course need to figure out what's up with the tests, and think about how to document this well.

File: gabbi/tests/gabbits_intercept/backref.yaml [hunk 1]
```diff
@@ -73,3 +73,13 @@ tests:
 - name: backref json fail end
   xfail: true
   url: $RESPONSE['url']
+
+- name: get a historical response
+  GET: $HISTORY['post some json'].$RESPONSE['a']
```
Comment (cdent) [inline gabbi/tests/gabbits_intercept/backref.yaml]: In py35 `urlunsplit` has slightly different behavior so in order for this test to work properly (and to have a correct url) this needs to be:

    GET: /$HISTORY['post some json'].$RESPONSE['a']

File: gabbi/tests/gabbits_intercept/backref.yaml [hunk 2]
```diff
@@ -73,3 +73,13 @@ tests:
...
+  response_headers:
+      x-gabbi-url: $SCHEME://$NETLOC/1
+
+- name: get a historical response via jsonpath
+  GET: $HISTORY['post some json'].$RESPONSE['$.b']
```
Comment (cdent) [inline gabbi/tests/gabbits_intercept/backref.yaml]: and here

**Comment by samuelfekete** [comment]
@cdent, thanks for the quick response. I am not going away. Will do another commit soon to fix the failing tests.

**Comment by jd** [comment]
👍 for the feature. I am not competent enough to validate the code. but this is the way I envisioned the functionality for a long time.

**Comment by cdent** [comment]
I'll do some docs and release notes for this next week, unless someone gets there first. Thanks very much @samuelfekete

**Comment by samuelfekete** [comment]
Thanks @cdent

**Comment by cdent** [comment]
This has been released as `1.30.0`: https://pypi.python.org/pypi/gabbi/1.30.0

**Pull request merged**